Use 'true' and 'false' booleans for Like module.

coleshaw opened the pull request:
Be consistent with the rest of the code base.

<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Fixes #14277 

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Use 'true' / 'false' instead of 'on' / 'off' strings for radio button values in the Like module settings page.


#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* Changes internal behavior of an existing part of Jetpack.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to Jetpack -> Settings -> Modules -> Likes -> Configure.
* Right-click on the radio buttons for "On for all posts" and "Turned on per post". Inspect the elements in the DOM.
* Verify that the `values` for each `input` element are `true` or `false` instead of `on` and `off.
* Toggle the radio button. For each value, you can confirm in the Edit Post page that the "Show likes" checkbox (under the Jetpack menu) is checked or unchecked by default.
   * When "On for all posts" is selected, the "Show likes" box should be checked by default for existing or new posts.
   * When "Turned on per post" is selected, the "Show likes" box should be unchecked by default for existing or new posts.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Use `true` and `false` values for the radio buttons in the Likes module settings.

coleshaw committed:
Use 'true' and 'false' booleans for Like module.

Fix #14277.

Be consistent with the rest of the code base.
Use 'true' / 'false' instead of 'on' / 'off'
strings for radio button values in the Like
module settings page.

zinigor commented:
Thank you for taking this bug! We appreciate any help with fixing issues. I can see you have found the place where we're using `on` and `off` strings instead of booleans.

This is a good start, but I have several suggestions, I'm going to add them as inline comments.

zinigor commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -599,13 +599,13 @@ function admin_settings_callback() {
 		// Checked (enabled)
 		switch( $new_state ) {
-			case 'off' :
+			case 'false':
```
Here you're changing the value from one string to another string, but the intention was to change that value to a boolean. Same with all the other values down there. Let's use `true` and `false` instead.

Plus, more importantly, on line 593 and 596 there's a default setting that still sets the variables to `on` if the value is not defined.

zinigor reviewed (changes requested):
I have left a few comments in the code, please take a look. I'm sorry if I'm not able to respond quickly, things should clear up next week.

coleshaw commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -494,13 +494,13 @@ function admin_settings_init() {
 				<div>
 					<label>
-						<input type="radio" class="code" name="wpl_default" value="off" <?php checked( $this->is_enabled_sitewide(), false ); ?> />
+						<input type="radio" class="code" name="wpl_default" value="false" <?php checked( $this->is_enabled_sitewide(), false ); ?> />
```
Ah, sorry I misunderstood the issue! Let me take a closer look at your feedback and the other modules that were already updated.

coleshaw committed:
Revert to master state, with on / off.

coleshaw committed:
Convert to 1/0 in values and convert to booleans

[not verified]  (no nonces used for these forms)

coleshaw committed:
Use 1/0 instead of on/off in Likes module.

coleshaw committed:
Account for WP.com Likes value.

When toggling the "Someone likes one of my posts."
value on WP.com's Discussion settings, it still sends
"on" and "off" values.

coleshaw committed:
Convert on/off in Subscriptions module to 1/0.

Need to account for WP.com toggle values of
"on" and "off" sent for "Someone follows my blog",
in addition to the 1/0 values set locally.

coleshaw committed:
Revert 1/0 instead of on/off in Likes and Subscriptions

This reverts commit 01fa1fd650644197753a1662eb3ef0cff09a7839.

coleshaw commented:
@zinigor , are these changes more in line with the original intent of the issue?

Question: I was going to change the `on/off` values to `1/0` in `modules/likes.php` `\Jetpack_Likes::admin_discussion_likes_settings_validate`, which I think was the original issue you posted about. I have it working on the site's Discussion Settings page with `1/0`. However, if you toggle the values for "Someone likes one of my posts" and "Someone follows my blog" in your WP.com dashboard, it seems like the option values get set to `on/off` instead of `1/0`. Does that mean:

* WP.com sends `on/off` to my site, when I toggle those values on WP.com?
* Is there a filter somewhere, where I can change that behavior? OR, should I just account for either `1/0` or `on/off` being the current option value?

Finally, I'm not really sure I should be fixing the code climate issue that was flagged (duplicate-looking code in the Likes module). Is there something I should do about that?

Thanks!

zinigor commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -590,22 +590,22 @@ function is_comments_enabled() {
 		$db_state  = $this->is_enabled_sitewide();
 
-		$reblogs_new_state = ! empty( $_POST['jetpack_reblogs_enabled'] ) ? $_POST['jetpack_reblogs_enabled'] : 'on';
+		$reblogs_new_state = ! empty( $_POST['jetpack_reblogs_enabled'] ) ? $_POST['jetpack_reblogs_enabled'] : false;
```
I'm sorry, I don't understand why you are changing the default value from positive to negative? Should this be `true` instead?

coleshaw replied:
Yes, this was kind of confusing to me ... I will try to explain. Since I changed the input values to `"0"` and `"1"` on lines 497 and 503, when the option is disabled, [`empty("0")` evaluates to `true`](https://www.php.net/manual/en/function.empty.php) and thus uses the `false` / `else` side of the ternary -- this is no longer a default value for when no input value is passed in, but rather the value when "0" is passed in (disabled). When the values were `on` and `off`, those strings were never considered `empty` and so the `else` side of the ternary was actually a default value.

If I put `true` there, it is impossible to disable the option, it always appears as enabled on page load.

coleshaw replied:
Note that on line 635 there is similar behavior (I think??), where the default value for the ternary is `false`... I assume 0 / 1 are being passed in there.

zinigor commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -590,22 +590,22 @@ function is_comments_enabled() {
 		// Checked (enabled)
-		switch( $new_state ) {
-			case 'off' :
+		switch ( (bool) $new_state ) {
```
This will not work the way you expect it to. You can try it yourself with a snippet like this:
```
<?php
echo (bool) 'off';
```
You'll see that a string `off` is typecast as `true` instead of `false`.

coleshaw replied:
Hm...since I changed the values to `0` and `1` in the inputs (lines 497 / 503 and 517 / 523), wouldn't casting those as boolean result in `false` / `true`?

zinigor commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -590,22 +590,22 @@ function is_comments_enabled() {
 				update_option( 'disabled_likes', 1 );
 				break;
-			case 'on'  :
+			case true:
```
I'd leave both cases here instead, as well as in the case with `false`. Let's leave it like:
```
case 'on':
case true:
```

zinigor commented on modules/likes/jetpack-likes-settings.php:
```diff
@@ -614,14 +614,14 @@ function admin_settings_callback() {
 
-		switch( $reblogs_new_state ) {
-			case 'off' :
+		switch ( (bool) $reblogs_new_state ) {
```
Same thing here, let's avoid typecasting the value directly.

zinigor reviewed (changes requested):
Thanks for working on this! We're getting closer, but this still needs some work.

coleshaw commented:
Sorry for all my confusion, I feel like I've muddied up the waters quite a bit by not having a good grasp of the original issue. I appreciate you working with me through the details ... if I've gone too far off track, please let me know.
